---
title: "RNA-seq workshop"
subtitle: "quantification and differential expression analysis"
author: "Yue Jiang"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r prepare_table, include=FALSE, eval=FALSE}
ctrl_A <- read.table("counts/GSM1435022_YJMi4counts.txt", header=FALSE, sep="\t", stringsAsFactors=FALSE)
ctrl_B <- read.table("counts/GSM1435024_YJ1counts.txt", header=FALSE, sep="\t", stringsAsFactors=FALSE)
ctrl_C <- read.table("counts/GSM1435026_YJ3counts.txt", header=FALSE, sep="\t", stringsAsFactors=FALSE)
stim_A <- read.table("counts/GSM1435023_YJMi5counts.txt", header=FALSE, sep="\t", stringsAsFactors=FALSE)
stim_B <- read.table("counts/GSM1435025_YJ2counts.txt", header=FALSE, sep="\t", stringsAsFactors=FALSE)
stim_C <- read.table("counts/GSM1435027_YJ6counts.txt", header=FALSE, sep="\t", stringsAsFactors=FALSE)

count_table <- data.frame(Gene=ctrl_A[, 1],
                          Ctrl_A=round(ctrl_A[, 21]),
                          Ctrl_B=round(ctrl_B[, 21]),
                          Ctrl_C=round(ctrl_C[, 21]),
                          Stim_A=round(stim_A[, 21]),
                          Stim_B=round(stim_B[, 21]),
                          Stim_C=round(stim_C[, 21]))

write.table(count_table, "counts/count_table.tab", sep="\t", row.names=FALSE)
```

## Set up environment
Let's set up the environment where we work. Let's assume that you have a folder on your desktop called `achems`, and we will use that folder as our working directory. We will download our input file (the count table) to that folder.

If a library hasn't been installed yet, you can install it using e.g. `install.packages("ggplot2")`. For bioconductor libraries, use `biocLite` instead, e.g.
```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
```

- set working directory
- download count table
- load libraries
- custom plot theme

```{r setup, message=FALSE, warning=FALSE}
setwd("~/Desktop/achems/")
system("curl -O https://raw.githubusercontent.com/Yue-Jiang/achems/master/count_table.tab")

library(ggplot2)
library(ggrepel)
library(dplyr)
library(tidyr)
library(DESeq2)
library(grid)
library(gridExtra)
library(knitr)
library(superheat)
library(plotly)
library(clusterProfiler)
library(org.Mm.eg.db)

# just a theme for plotting! don't worry about it!
my_theme <- function(base_size=14) {
  (theme_bw(base_size = base_size)
    + theme(plot.title = element_text(face = "bold",
                                      size = rel(1.2), hjust = 0.5),
            text = element_text(),
            panel.background = element_rect(colour = NA),
            plot.background = element_rect(colour = NA),
            panel.border = element_rect(colour = NA),
            axis.title = element_text(face = "bold",size = rel(1)),
            axis.title.y = element_text(angle=90,vjust =2),
            axis.title.x = element_text(vjust = -0.2),
            axis.text = element_text(), 
            axis.ticks = element_line(),
            panel.grid.major = element_line(colour="#f0f0f0"),
            panel.grid.minor = element_blank(),
            legend.key = element_rect(colour = NA)
    )
  )
}
```

## Read in count table
Differential expression analysis softwares usually ask that you have a table of your expression data. Each software may have slightly different requirements in the way they want their input data prepared. For DESeq2, we need to have a data frame with expression levels in counts (integer).

- Each row is a gene
- Each column is a sample

```{r load_data}
count_df <- read.table("count_table.tab", sep="\t", header=TRUE, stringsAsFactors=FALSE)
```

This is what the count table looks like (showing top 10 rows, 23481 rows total).
```{r, echo=FALSE}
kable(count_df[1:10, ])
```

## Make your design table
Design table specifies experiment conditions for your samples. In this experiment, we have 6 samples with treatment (control vs stimulated) and littermate information (A, B, C). We are interested in genes differentially expressed in stimulated samples as compared to the control.

```{r design}
design_df <- data.frame("Sample"=c("Ctrl_A", "Ctrl_B", "Ctrl_C", "Stim_A", "Stim_B", "Stim_C"),
                        "Treatment"=c("control", "control", "control", "stimulated", "stimulated", "stimulated"),
                        "Litter"=c("A", "B", "C", "A", "B", "C"))
```

The design table looks like this.
```{r, echo=FALSE}
kable(design_df)
```

## Differential expression analysis
Let's perform differential expression analysis using DESeq2. Remember that we are only interested in the effect of treatment. The litter effect is included in the model to absorb the variance explained by litter so that we get better sensitivity. It would be slightly better to use EdgeR, where a mixed effect model is implemented, allowing for modeling litter as random effect, but we won't go into the details there.

- prepare input data as DESeq2 expects
- run DESeq2 following the manual

```{r de, cache=TRUE, message=FALSE, warning=FALSE}
countData <- as.matrix(count_df[, -1])
rownames(countData) <- count_df$Gene
colData <- design_df
rownames(colData) <- colData$Sample
# make sure sample orders are the same
countData <- countData[, rownames(colData)]

dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ Treatment + Litter)
# filter out genes lowly expressed, this is an arbitrary cutoff
dds <- dds[ rowSums(counts(dds)) > 5, ]
dds <- DESeq(dds)

de <- results(dds, contrast=c("Treatment", "stimulated", "control"))
```

Then we get the DE result. Here are the top 10 most DE genes.
```{r top_de}
kable(de[order(de$padj)[1:10], ])
```

## Visualizing DE result
For clarity, let's make a data frame to hold the data for plotting. The data frame will contain the following columns:

- Gene name
- base mean
- log2 fold change
- FDR corrected p-value
- An annotation column indicating whether this gene is an OR
- Another annotation column indicating whether this gene is differentially expressed

In the original paper, the FDR correction is performed on odorant receptor genes only. For simplicity we skip that.

```{r de_df}
de_df <- data.frame("Gene"=rownames(de),
                    "baseMean"=de$baseMean,
                    "log2FoldChange"=de$log2FoldChange,
                    "pAdj"=de$padj) %>%
  # fill N/A adjusted p-values with 1's
  replace_na(list(pAdj=1)) %>%
  # add an annotation column indicating whether this gene is an OR
  mutate(Annotation=case_when(grepl("^Olfr", .$Gene) ~ "Odorant receptor",
                              TRUE ~ "Other genes")) %>%
  # add another annotation column indicating whether gene is enriched / decreased / not DE
  mutate(DE=case_when(.$pAdj <= 0.05 & .$log2FoldChange > 0 ~ "Enriched (p<=0.05)",
                      .$pAdj <= 0.05 & .$log2FoldChange < 0 ~ "Decreased (p<=0.05)",
                      TRUE ~ "p>0.05"))
```

### MA plot
An MA plot is plotting the log2 fold change vs mean expression level.
We can customize the MA plot by highlighting genes of interest (for example, all odorant receptor genes).

```{r de_ma, fig.width=6, fig.height=6}
p1 <- ggplot(de_df, aes(baseMean, log2FoldChange, color=DE)) +
  geom_point(alpha=0.8, size=1) +
  scale_x_log10() +
  scale_color_manual(values=c("forestgreen", "red", "grey80")) +
  my_theme()
p2 <- ggplot(de_df %>% arrange(desc(Annotation)), aes(baseMean, log2FoldChange, color=Annotation)) +
  geom_point(alpha=0.8, size=1) +
  scale_x_log10() +
  scale_color_manual(values=c("red", "grey80")) +
  my_theme()
grid.arrange(p1, p2, ncol=1)
```

### Volcano plot
Another popular way to visualize DE analysis is to plot log scaled p-values against log scaled fold change. We can customize the volcano plot by labeling the ORs that are enriched in stimulated samples.

```{r de_volcano, fig.width=6, fig.height=8}
ggplot(de_df, aes(log2FoldChange, log10(pAdj), color=DE)) +
  geom_point(alpha=0.8, size=1) +
  geom_text_repel(data=de_df %>% filter(Annotation == "Odorant receptor" & DE == "Enriched (p<=0.05)"), aes(label=Gene), segment.colour="grey50", color="grey10") +
  scale_y_reverse() +
  scale_color_manual(values=c("forestgreen", "red", "grey80")) +
  my_theme()
```

## Gene ontology
Are certain pathways enriched in the DE genes?

```{r go, cache=TRUE}
de_genes <- as.character(filter(de_df, pAdj <= 0.05)$Gene)
background_genes <- as.character(filter(de_df, baseMean >= 50)$Gene)
go <- enrichGO(gene=de_genes,
               universe=background_genes,
               OrgDb = org.Mm.eg.db, keytype = 'SYMBOL',
               ont = "BP",
               pAdjustMethod = "BH",
               pvalueCutoff = 0.1,
               qvalueCutoff = 0.1
               )
kable(go@result)
```

Since there are some redundancy in the GO terms, we can simply them.
```{r sim_go, cache=TRUE}
sim_go <- simplify(go, cutoff = 0.7, by = "p.adjust", select_fun = min, measure = "Rel")
kable(sim_go@result)
```

## Clustering
Sometimes we want to do exploratory analysis by clustering samples based on RNA expression profiles and see if the clustering makes biological sense. Let's see a few examples.

### hierarchical clustering
Let's cluster the samples based on OR expression. Well, they actually cluster by litter... So it's important to account for litter effect as we did before!

```{r clustering}
orCountData <- countData[grepl("^Olfr", rownames(countData)) & rowSums(countData) > 5, ]
superheat::superheat(t(log1p(orCountData)),
                     pretty.order.rows = TRUE,
                     pretty.order.cols = TRUE,
                     left.label.size = 0.3,
                     left.label.text.size = 3,
                     left.label.text.alignment = "right",
                     left.label.col = "white",
                     # scale the matrix columns
                     scale = TRUE,
                     row.dendrogram = TRUE,
                     col.dendrogram = TRUE,
                     legend=FALSE)
```

### PCA
Just to showcase some PCA visualization. 
```{r pca}
or_pca <- data.frame(prcomp(t(log1p(orCountData)), scale.=TRUE)$x)
plot_ly(or_pca, x = ~PC1, y = ~PC2, z= ~PC3, text = rownames(or_pca)) %>%
  add_markers() %>%
  add_text() %>%
  layout(showlegend = FALSE)
```